### ISSUE TITLE
Functional testing & browser automation

### ISSUE THREAD

**Issue by virtualtam** [shaarli/Shaarli#342]
Relates to #71 (unit testing) & #130 (CI discussion)
Follow-up to #16 (API / programmable access)

### Original thread
(from #16)

> @mro:
> I consider the http get parameters, html form content and page flow the API.
> - How could we - as a part of the travis build - ensure a certain behaviour?
> - Would bash+curl test scripts make sense?
> - How could phpunit help here?

> @virtualtam:
> PHPUnit itself is not designed to run functional tests, as you need to:
> - run an HTTP server
> - run an HTTP client to send, receive and verify requests
>
> This is rather the domain of [SeleniumHQ](http://docs.seleniumhq.org/), which can be used both for verifying:
> - HTTP headers
> - generated pages
>
> Some info:
> - http://docs.travis-ci.com/user/gui-and-headless-browsers/
> - https://www.browserstack.com/automate/continuous-integration
>
> Any feedback/PR on using Selenium for PHP browser automation is more than welcome, this would be great to have for [v0.8.0](https://github.com/shaarli/Shaarli/milestones/0.8.0) to help with templating and W3C compliance checks

> @mro:
> Looks a bit hairy but may be worth while: http://docs.travis-ci.com/user/languages/php/#Apache-%2B-PHP
>
> test multiple shaarli versions at once: https://github.com/mro/phptestdemo
>
> The tests mentioned above show a strange pattern for 'wrong password' login failures with certain, mostly recent, releases: https://travis-ci.org/mro/phptestdemo/builds/78849088
>
> I suspect php caching which I'm clueless about in travis environments.
>
> Any ideas?

### Test repository / Travis scripts
[mro/phptestdemo](https://github.com/mro/phptestdemo) - [Travis builds](https://travis-ci.org/mro/phptestdemo/branches)

**Comment by virtualtam** [comment]
> 'wrong password' login failures

Login issues are currently being addressed in #338, which may imply that Shaarli releases `v0.5.{0,1,2,3}` are deemed to fail on login / auth

> I suspect php caching which I'm clueless about in travis environments.
> Any ideas?

Travis uses VMs (legacy infrastructure) and Docker containers. I'd go for first running tests on pure Docker containers, then see if it can be transposed to Travis builds.

There are two base Docker images as well as some documentation available in [shaarli/docker-shaarli](https://github.com/shaarli/docker-shaarli), that can be inherited and tweaked for testing purposes, let me know if you need help with it :)

Regarding [phptestdemo](https://github.com/mro/phptestdemo), it'd be very helpful to have a brief description of what's being deployed, which are the expected results, track issues & progress, etc.

Once stabilized, the relevant bits could make their way to the community documentation / repos, so that development efforts are centralized (better visibility, eases/encourages contribution)

**Comment by mro** [comment]
tl;dr: I care about tests because of https://github.com/mro/ShaarliOS/issues/1.

Writing integration/functional tests for Shaarli is purely selfish - I need to know what backends work with https://github.com/mro/ShaarliOS. So the repo https://github.com/mro/phptestdemo mentioned above may change name or disappear completely and testing may become part of ShaarliOS.

So my focus is a very limited feature set (post a link) but with a large range of shaarli (and php) versions. Also I don't care about unit tests - I only care what's going over the wire. Maybe https://github.com/dimtion/Shaarlier has a similar incentive and may join in.

About Docker:

The quick builds are nice, but: As https://github.com/sebsauvage/Shaarli (vanilla shaarli I call it) is the baseline that I want to support with https://github.com/mro/ShaarliOS - and vanilla requires port 80 to run afaik, and travis + docker currently isn't easily serving the 'sudo' ports < 1024 - AND it's bitching a bit (https://travis-ci.org/mro/phptestdemo/builds/78847294) - I won't push docker based tests any further for now.

**Comment by virtualtam** [comment]
> Writing integration/functional tests for Shaarli is purely selfish

Yet community Shaarli could benefit from any effort made towards it :)

> tl;dr: I care about tests because of mro/ShaarliOS#1.
> Ideally there should be a common agreement on the 'API' between vanilla shaarli, community shaarli and ShaarliOS or other clients.

There could be a connector for legacy (development stalled), and an API for the community version (active development). Headless interaction with Shaarli _should not be_ dependent on the generated HTML (else, there's going to be trouble when the Plugin & Template Management Systems are merged)

> Explaining the Tests [...] Either they're simple to the reader or it may be a waste of time anyway

My 2 cents as a software release manager / CI admin:

The code being crystal-clear does not mean that it is useless to add brief comments; you never know if the person who's interested in your code:
- is experienced in that particular language
- wants to delve into the code or rather have an overall idea of what a script / class / method does
- is even browsing your source code, or either using a generated documentation or a cross-reference engine (e.g. [OpenGrok](http://opengrok.github.io/OpenGrok/)) 

> About Docker

Though Travis + Docker come with some limitations, it'd be quite easy to setup a custom CI server to:
- run web servers on:
    - `:80` for legacy Shaarli
   - `:80` + other ports for the latest community Shaarli
- run tests on a given Shaarli / PHP matrix; Docker would be the best choice here in terms of development efforts, thanks to template inheritance:
    - base template for each supported PHP version
    - inherited/parametrized template for each supported Shaarli version
- display more relevant / up-to-date compatibility information on ShaarliOS' status page, e.g. a badge per tested PHP / Shaarli version

Such a platform could be deployed quite easily using e.g. Jenkins CI.

AFAIK, users that have switched to the community Shaarli keep it up-to-date, hence active feedback on `master`; I'd tend to second @dimtion's comment on mro/ShaarliOS/issues/2 in that there's no support for legacy Shaarli, so it'd be worth migrating.

**Comment by mro** [comment]
>> Writing integration/functional tests for Shaarli is purely selfish

> Yet community Shaarli could benefit from any effort made towards it :)

indeed.

> Headless interaction with Shaarli should not be dependent on the generated HTML (else, there's going to be trouble when the Plugin & Template Management Systems are merged)

agreed, but `<form><input …>` is an example of what I mean - it's html markup and produces human readable display items but at the same time is machine readable API - the POST parameters. A little different are the ['is logged in' detection](https://github.com/mro/ShaarliOS/blob/8ec1568edceadae0d6b0a31e834ba2bba4074ffb/ios/ShaarliOS/ShaarliM.m#L200) or the (optional) [shaarli title](https://github.com/mro/ShaarliOS/blob/8ec1568edceadae0d6b0a31e834ba2bba4074ffb/ios/ShaarliOS/ShaarliM.m#L221).

I think about using [this xslt transformation](https://github.com/mro/phptestdemo/blob/f205057884a43cc033805c458c75ab49c49c9920/scripts/response.xslt) to destill the html into a compact xml form ([RelaxNG schema t.b.d.](http://blog.mro.name/2010/05/xml-toolbox-relax-ng-trang/)) which

1. is much simpler to parse from [C (libxml2)](https://github.com/mro/ShaarliOS/blob/8ec1568edceadae0d6b0a31e834ba2bba4074ffb/ios/ShaarliOS/ShaarliM.m#L168),
2. can serve as a testbed for template creators what structural markup criteria are relevant,
3. may even be provided - and customized - by the template authors (or fallback to the one above).

> The code being crystal-clear does not mean that it is useless to add brief comments

...remains the question how brief *brief* is :-)

But seriously, I disagree with the language skill part (there's usually tutorials out there to be linked to), agree with coarse grained explanations and want to add explanations to my future self. I usually add lots of inline code comments what I used and read while coding.

However, I won't neither explain what a `.travis.yml` is, but add a link to travis docs as a first line, nor explain what a `scripts/run-test.sh` or `scripts/test-post.sh` might do.

Enough spoken - pull requests welcome :-)

P.S.: There's one thing I'd appreciate help from you @virtualtam - do you have an idea why the [Atom Feed doesn't keep the original default entry after adding the first link](https://travis-ci.org/mro/phptestdemo/jobs/78893126#L363)?

P.P.S.: Each test can be run in isolation when the shell environment variables `USERNAME`, `PASSWORD` and `BASE_URL` are set accordingly

**Comment by mro** [comment]
not much happening here – is this still a serious goal?

**Comment by nodiscc** [comment]
Hi, I am confused/curious about this issue's goals. Login failures where mentioned but I'm not sure what there is to test.

> How could we [...] ensure a certain behaviour?

can someone give an example of such an expected behaviour?

**Comment by mro** [comment]
goals: Testing full use cases (click paths) and testing reliable (compatible) interfaces on a machine understandable level.

E.g. expecting certain reponse status codes or existence of tags/ids (or even lexical text structures as in Javascript error messages). They're invisible to humans, but relevant for non-human callers.

E.g. #356 or #533. Relevant also in context of #16 or #113, https://github.com/mro/ShaarliOS/issues/1

But if there's no action intended in the near future, I'd advocate to close this issue and re-raise it again when there's a serious intent to do something about it. Currently it IMO just clutters the issue list and distracts from the *real* issues.

**Comment by virtualtam** [comment]
Yes, this topic is still worth addressing ;-)

@nodiscc where unitary tests ensure the code works well on the server side, functional/browser testing rathers covers how usable a GUI is by users, and may highlight design/usability flaws.

A similar tool is [Android's UI Automator](http://developer.android.com/training/testing/ui-testing/index.html), that allows testing the behaviour of a mobile app over several device configurations (e.g. different screen sizes and input peripherals).

> But if there's no action intended in the near future, I'd advocate to close this issue and re-raise it again when there's a serious intent to do something about it. Currently it IMO just clutters the issue list and distracts from the real issues.

Agreed, this issue is rather a generic discussion that can now be archived ;-)
